WIP: Authenticator taxonomy (replaces #842)

emlun opened the pull request:
This replaces PR #842. PR #842 was from a branch in https://github.com/emlun/, which means that PRs into #842 end up in https://github.com/emlun/ and don't get previews and diffs rendered. This PR is from a branch in this repository, so the bots can do those things. Sorry for the noise...

Original post from #842 below.

---

_This is a work in progress._

This aims to resolve #422.

This is what I've come up with so far. It will likely need some rather major surgery before it's ready to be merged, so I'd be happy for both detail-level corrections and high-level restructuring suggestions. Other editors are welcome to push commits directly into this PR, too.

Some issues I've identified while writing this:

- I think we've implicitly assumed throughout the spec that authenticators will always require user verification to create and use client-side-resident credential private keys, but this doesn't seem to be documented in the spec. CTAP2 also doesn't seem to specify this behaviour. The "username-less use case" I've written in here is probably not very useful and would be merged into the "single-step use case" given the above requirement, but without that requirement it remains a possible scenario.
  - [ ] Resolution:
- Is there someplace we can refer to for "authentication factor" and the related terms (known/possessed/biometric factor) instead of defining them in the spec? The Internet Security Glossary (RFC 4949) doesn't seem to have them.
  - [x] Resolution: Yes, [NIST SP800-63-3](https://pages.nist.gov/800-63-3/sp800-63-3.html#af).
- This text might not belong in the Authenticator Model section.
  - [ ] Resolution:


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/emlun/webauthn/pull/842.html" title="Last updated on Jun 14, 2018, 8:15 PM GMT (4fcb566)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/842/4fd5dd5...emlun:4fcb566.html" title="Last updated on Jun 14, 2018, 8:15 PM GMT (4fcb566)">Diff</a>

emlun commented:
Sorry for the noise. I replaced #842 with this so I could make proper mini-PRs into this one, since #842 was from a branch in https://github.com/emlun/webauthn which made things difficult.

@selfissued You reviewed and [approved][approve] commit 75f348ee38dc8f4311d8e0aab191934a43d69a7e in #842, and that commit is in this PR as well. The commits that have been added since then are ef272ad through 5945017 (`git diff 75f348 5945017`). You can see the total changes from those commits [here][new-diff].

[approve]: https://github.com/w3c/webauthn/pull/842#pullrequestreview-126205207
[new-diff]: https://github.com/w3c/webauthn/pull/956/files/75f348ee38dc8f4311d8e0aab191934a43d69a7e..5945017a25332720f232eb2cfc83f104df1768a9

emlun commented:
The review https://github.com/w3c/webauthn/pull/842#pullrequestreview-113755127 also pertains to this, and much of it is still outstanding.

equalsJeffH commented:
@emlun wrote: 
> The review https://github.com/w3c/webauthn/pull/842#pullrequestreview-113755127 also pertains to this, and much of it is still outstanding.

Ok, thx, I'll wait to re-review until you say so.

emlun commented:
I think most of the review https://github.com/w3c/webauthn/pull/842#pullrequestreview-113755127 has now been accounted for. Remaining issues:

- Do we keep the figure? It's not very interesting, as discussed in https://github.com/w3c/webauthn/pull/842#discussion_r195583403 and https://github.com/w3c/webauthn/pull/842#discussion_r196526039; and arguably incomplete, as noted in https://github.com/w3c/webauthn/pull/842#discussion_r182870904.
- I don't quite know how to go about https://github.com/w3c/webauthn/pull/842#discussion_r195888019; awaiting guidance from @equalsJeffH.

emlun commented:
@equalsJeffH @selfissued I broke the use cases section out into PR #979, so this PR is now a lot smaller. The two remaining issues in [my previous comment](https://github.com/w3c/webauthn/pull/956#issuecomment-400418026) now pertain to #979 instead, so I think I have addressed all review comments on what now remains in this PR.

emlun commented:
@selfissued I also un-rewrapped the line breaks for your reviewing pleasure. :)

equalsJeffH commented on index.bs:
```diff
@@ -2416,6 +2427,47 @@ This is because the first 37 bytes of the signed data in a FIDO U2F authenticati
+Issue: Expand this section with further discussion about the other dimensions mentioned above.
+
+
+### <dfn>Attachment Modality</dfn> ### {#sctn-authenticator-attachment-modality}
```
I'm inclined to fully-qualify this term, especially in the heading here, as "Authenticator Attachment Modality".

emlun replied:
Fixed in #979.

equalsJeffH commented on index.bs:
```diff
@@ -2416,6 +2427,47 @@ This is because the first 37 bytes of the signed data in a FIDO U2F authenticati
+
+Some [=platform authenticators=] could possibly also act as [=roaming authenticators=] depending on context. For example, a
+[=platform authenticator=] integrated into a mobile device could make itself available as a [=roaming authenticator=] via
+Bluetooth. In this case the [=client=] would identify it only as a [=roaming authenticator=], and not as a [=platform
```
suggest:
identify it only -> recognize it

emlun replied:
Done.

equalsJeffH commented on index.bs:
```diff
@@ -2416,6 +2427,47 @@ This is because the first 37 bytes of the signed data in a FIDO U2F authenticati
+Bluetooth. In this case the [=client=] would identify it only as a [=roaming authenticator=], and not as a [=platform
+authenticator=].
+
+The primary use case for [=platform authenticators=] is to register a particular [=client device=] as a "trusted device" available
```
The primary -> A primary  ?

emlun replied:
Done.

equalsJeffH commented on index.bs:
```diff
@@ -2416,6 +2427,47 @@ This is because the first 37 bytes of the signed data in a FIDO U2F authenticati
+of not needing a [=roaming authenticator=] for future [=authentication ceremonies=], e.g., the user will not have to dig around in
+their pocket for their key fob or phone.
+
+The primary use case for [=roaming authenticators=] is for initial [=authentication=] on a new [=client device=], or on [=client
```
The primary -> A primary ?

emlun replied:
Done.

nadalin commented:
@selfissued Please review
